### PR TITLE
feat: add telemetry to companion

### PR DESCRIPTION
**Pull request by SgtPooki**
## Things implemented:

- [x] View tracking of companion pages
- [x] Session tracking (powering down the extension ends your session and stops tracking sessions)
- [x] consent management: disabling telemetry (opting-out) prevents the above two things from happening, but we do send an event indicating that you've opted-out

## Things to do
- [x] #1114
- [x] Fix CI failures
- [x] Need to discuss with @BigLep and @tinytb w.r.t. their discussion on 2023-01-05 whether #1114 needs to be a pre-req for merging this

image

### CODE REVIEW

**Comment by SgtPooki** [comment]
@juliaxbow I would love your opinion on the formatting of the telemetry section

**Comment by juliaxbow** [comment]
**Descriptions**
- **Telemetry section** In the description, would say "To assist in our efforts, please consent to the collection of these metrics by toggling "on" in the corresponding section. You can change this setting at any time by navigating back to the Companion Preferences page."  It's a little unusual to offer consent via a toggle but I think it's ok?
- Is the toggle default to on? If so, we should change the wording even further. Something like "IPFS Companion is collecting minimal telemetry data to improve and prioritize our work. You may opt out of the collection of these metrics by toggling "off" in the corresponding section below. You can change this setting ......"
- **Metrics Section** the metric collection description, I would change wording to "Sends ‘session’ and ‘view’ events only when ipfs-companion is enabled and your browser is not idle." Current wording raises the question of who is sending this information where? 
- The section header "minimal metrics: ..." makes it seem like there are additional, more intrusive metrics somewhere. Until we have other metrics we're tracking I think we can just call it "Session & Companion views." The Telemetry description already points to the fact that these are minimal. 

**Other**
- "Read More" links on the page have a link-out icon, so would add that in. (in the screenshot below i have the same link blue because I think all link-out icons should look like links but keep consistent with the rest of the page for now ;) )
- The toggle is centered but it doesn't look like it is. Until we have more sections & metrics that we're tracking, I would move it up to be top-aligned with the first sentence in the description. I know this is inconsistent with all other toggles on the page but to me it's more clear what the user is agreeing to. This is subjective, though, so feel free to ignore.

Incl screenshot to show the _suggested_ toggle alignment
Screen Shot 2022-12-16 at 8 07 57 PM

File: add-on/_locales/en/messages.json [hunk 1]
```diff
@@ -726,5 +726,53 @@
...
+  "option_telemetryGroupMinimal_title": {
+    "message": "Minimal metrics: session & companion views",
+    "description": "A title for the 'minimal' grouping of metrics we collect (option_telemetryGroupMinimal_title)"
+  },
+  "option_telemetryGroupMinimal_description": {
+    "message": "We send `session` and `view` events only if ipfs-companion is enabled and your browser is not idle.",
+    "description": "A description for the 'minimal' grouping of metrics we collect (option_telemetryGroupMinimal_description)"
+  },
+  "option_telemetryGroupMinimal_session_description": {
+    "message": "A `session` event - every 60 seconds.",
```
Comment (lidel) [inline add-on/_locales/en/messages.json:747]: Would this be better?
```suggestion
    "message": "A `session` event  for every minute the extension was active.",
```

File: add-on/_locales/en/messages.json [hunk 2]
```diff
@@ -726,5 +726,53 @@
...
+    "description": "An explanation of the `session` telemetry we collect (option_telemetryGroupMinimal_session_description)"
+  },
+  "option_telemetryGroupMinimal_view_description": {
+    "message": "A `view` event - only for ipfs-companion pages.",
```
Comment (lidel) [inline add-on/_locales/en/messages.json:751]: Users have no idea browser extensions are regular webpages behind the scenes.
Maybe more explicit:

```suggestion
    "message": "A `view` event for every time a part of extension's user interface is visible.",
```

File: add-on/src/lib/options.js [hunk 3]
```diff
@@ -31,7 +33,11 @@ export const optionDefaults = Object.freeze({
   importDir: '/ipfs-companion-imports/%Y-%M-%D_%h%m%s/',
   useLatestWebUI: false,
   dismissedUpdate: null,
-  openViaWebUI: true
+  openViaWebUI: true,
+  telemetryGroupMinimal: true,
+  telemetryGroupMarketing: false,
+  telemetryGroupPerformance: false,
+  telemetryGroupTracking: false
```
Comment (lidel) [inline add-on/src/lib/options.js:40]: - rename `telemetryGroupMinimal` to `telemetryFeatures`
- remove `telemetryGroupMarketing` and `telemetryGroupTracking` (we are not doing these, afaik don't plan to do it, so no reason to trigger anyone with this language)

File: add-on/src/lib/telemetry.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  const enabledConsentGroups = Object.keys(obj).filter(key => obj[key] === true)
+  console.log('enabledConsentGroups: ', enabledConsentGroups)
+  return enabledConsentGroups
+}
+function logConsent () {
+  console.log('checkConsent(\'minimal\'): ', getMetricsProviderInstance().checkConsent('minimal'))
+  console.log('checkConsent(\'marketing\'): ', getMetricsProviderInstance().checkConsent('marketing'))
+  console.log('checkConsent(\'performance\'): ', getMetricsProviderInstance().checkConsent('performance'))
+  console.log('checkConsent(\'tracking\'): ', getMetricsProviderInstance().checkConsent('tracking'))
```
Comment (lidel) [inline add-on/src/lib/telemetry.js:34]: Remove all `console.log` calls, use scoped `debug` logger instead – see `src/lib/on-installed.js`, we want something similar:

```js
import debug from 'debug'
const log = debug('ipfs-companion:telemetry')
log(`checkConsent(foo): ${bar}`)
```
```

**Review by lidel — CHANGES_REQUESTED** [review]
Looks sensible, as long we address below:
1. We should not be sending requests to Countly every minute a user uses a browser. 
   (I assume that is what is happening? I was unable to build this and test locally)
   - This causes many issues, for example drains laptop battery by waking up WIFI when no internet conneciton would be needed otherwise. 
   - As discussed, figure a way to batch things so we report them only once an hour or once a day.  
1. Regular people don't know hat 'session' and 'view' means. 
   - See my suggestions below on humanizing the language :)

File: add-on/src/lib/notifier.js [hunk 5]
```diff
@@ -21,8 +21,8 @@ export default function createNotifier (getState) {
         return await browser.notifications.create({
           type: 'basic',
           iconUrl: browser.runtime.getURL('icons/ipfs-logo-on.svg'),
-          title: title,
-          message: message
+          title,
+          message
```
Comment (SgtPooki) [inline add-on/src/lib/notifier.js:25]: Lint fixes

File: add-on/src/lib/options.js [hunk 6]
```diff
@@ -1,8 +1,11 @@
 'use strict'
 
-import { isIPv4, isIPv6 } from 'is-ip'
 import isFQDN from 'is-fqdn'
+import { isIPv4, isIPv6 } from 'is-ip'
```
Comment (SgtPooki) [inline add-on/src/lib/options.js:4]: Vscode ordered imports on save. Willing to remove

File: add-on/src/options/page.js [hunk 7]
```diff
@@ -102,6 +103,13 @@ export default function optionsPage (state, emit) {
     logNamespaces: state.options.logNamespaces,
     onOptionChange
   })}
+  ${telemetryForm({
+    telemetryGroupMinimal: state.options.telemetryGroupMinimal,
+    telemetryGroupMarketing: state.options.telemetryGroupMarketing,
+    telemetryGroupPerformance: state.options.telemetryGroupPerformance,
+    telemetryGroupTracking: state.options.telemetryGroupTracking,
+    onOptionChange
+  })}
```
Comment (SgtPooki) [inline add-on/src/options/page.js:112]: @SgtPooki to update

File: add-on/src/types.d.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
+
```
Comment (SgtPooki) [inline add-on/src/types.d.ts:1]: Typescript types supported for linting. No work done for building validation (except from linter) nor exporting types (shouldn't be needed).

Errors show in vscode but don’t fail build yet

File: docs/telemetry/COLLECTED_DATA.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
+# Collected telemetry data
```
Comment (SgtPooki) [inline docs/telemetry/COLLECTED_DATA.md:1]: @whizzzkid @lidel first demo of what COLLECTED_DATA.md should look like as defined from ipfs-shipyard/ignite-metrics#35.  Lmk what you think

**Review by SgtPooki — COMMENTED** [review]
Self review

**Comment by SgtPooki** [comment]
> We should not be sending requests to Countly every minute a user uses a browser.
> (I assume that is what is happening? I was unable to build this and test locally)

I just tested again, and since the “autoTrack” is set to false, countly does not send session_duration events anymore (those are sent by countly if you call `.track_sessions()`). See my comment about state.active and starting/ending sessions.

Also, i feel like we should so some implementation of observability on “idle” state. Any thoughts on changing the current use of “runOfNotIdle” to a boolean on the state object?  We could have a single interval that checks browser.idle.queryState and updates state.isIdle. Then check that variable prior to actions.

We could also add “runIfNotIdle” as a function to the state object so it’s re-usable. 

The main thing i want to stay away from is a pattern of multiple intervals scattered throughout the codebase, so lmk if you two have any thoughts here please!

**Comment by SgtPooki** [comment]
@lidel @whizzzkid note that I pushed up a fix for the CI failures using the version of ignite-metrics at https://github.com/ipfs-shipyard/ignite-metrics/pull/60. will update package.json once I can use the version on npm again.

**Comment by SgtPooki** [comment]
FYI @lidel @whizzzkid I had some changes locally that weren't pushed up... getting those up now

**Comment by SgtPooki** [comment]
Agreed with @whizzzkid. Metrics are only sent (currently) when interacting with the extension (enabling/disabling, viewing different pages), so battery use should not be a concern.

**Pull request merged**